src/Iterators.jl

```diff
@@ -894,4 +895,94 @@ macro chain(ex)
     Expr(:let, Expr(:block, cycleex...), states...)
 end
 
+
+type MergedIter
+    it1   # iterator 1
+    it2   # iterator 2
+    lt    # custom `<` operator to compare elements of 2 iterators    
+end
+
+MergeIter(it1, it2) = MergeIter(it1, it2, isless)
+
+type MergedState
+    hd1   # current head of 1st iterator
+    hd2   # current head of 2nd iterator
+    s1    # current state of 1st iterator
+    s2    # current state of 2nd iterator
+end
+
+
+function start(it::MergedIter)
+    s1 = start(it.it1)
+    s2 = start(it.it2)
+    # doing 1 step ahead and caching next element of each iter
+    hd1, s1 = safenext(it.it1, s1)
+    hd2, s2 = safenext(it.it2, s2)
+    return MergedState(hd1, hd2, s1, s2)
+end
+
+"""
+Like `next`, but may be called on iterator that is already done. In this case
+new item is `nothing` and new state is the same as previous. 
+"""
+function safenext(it, s)
+    if done(it, s)
```

Comment:
True, for basic case an algorithm is pretty easy, but the devil is in the detail. To check that one of iterators is really done, we need not only test `done(it1, s1)`, but also see if `hd1` doesn't hold previously fetched element. If we do have one more element in `hd1`, we should take smaller from `hd1` and `hd2`. But second iterator may also be done, so we need to check if `hd2` isn't `nothing` too. So previous version of the `next()` function looked like this:

```
function Base.next(it::MergedIter, s::MergedState)
    if done(it, s) throw(ArgumentError("Iterator is done")) end
    it1 = it.it1; it2 = it.it2; hd1 = s.hd1; hd2 = s.hd2; s1 = s.s1; s2 = s.s2
    while hd1 == hd2 && !done(it2, s2)
        # move second iterator till heads are different or iterator is done
        hd2, s2 = next(it2, s2)
    end
    if done(it1, s1)
        # 1st iterator is done, but we have one saved value in hd1
        # TODO: hd2 == nothing
        if hd1 == nothing || hd1 > hd2
            next_hd2, next_s2 = safenext(it2, s2)
            return hd2, MergedState(hd1, next_hd2, s1, next_s2)
        else  # hd1 cached and hd1 < hd2 
            return hd1, MergedState(nothing, hd2, s1, s2)
        end
    elseif done(it2, s2)
        # 2st iterator is done, but we have one saved value in hd2
        if hd2 == nothing || hd2 > hd1
            next_hd1, next_s1 = safenext(it1, s1)
            return hd1, MergedState(next_hd1, hd2, next_s1, s2)
        else  # hd2 cached and hd2 < hd1
            return hd2, MergedState(hd1, nothing, s1, s2)
        end
    elseif hd1 < hd2
        next_hd1, next_s1 = safenext(it1, s1)
        return hd1, MergedState(next_hd1, hd2, next_s1, s2)        
    else
        next_hd2, next_s2 = safenext(it2, s2)
        return hd2, MergedState(hd1, next_hd2, s1, next_s2)
    end
end
```
Quite complicated, right? And it still doesn't handle all corner cases. 

It is also possible to unfold all cases into a single plain `if ... esleif ... elseif ... end` block, but given that you have at least 4 binary conditions (iterator 1 is done/not done, iterator 2 is done/not done, head 1 is empty/not empty, head 2 is empty/not empty) we get at least `2^4 = 16` blocks, each of which should be carefully thought out. And it doesn't even cover comparison of `hd1` and `hd2`. 

I don't really think that any of these 2 solutions would be easy to analyze or modify, so I've come up with an idea that it would be nice to wrap management of `(head, state)` into some kind of interface similar to normal iterators. In this MR I've done it using `safenext()` instead of `next()` and `complete()` instead of `done()`. This way we can express the whole algorithm into only 5 conditions (including head comparison):

```
if 1st iterator is completely done and no cached head is left
    return next element of the 2nd iterator
elseif 2nd iterator is done and no cached head is left
    return next element of the 1st iterator
elseif head 1 < head 2
    return head 1 and replenish it using next element of iterator 1 or `nothing`
elseif head 2 < head 1
    return head 2 and replenish it using next element of iterator 2 or `nothing`
elseif head 1 == head 2
    skip until they are different of we encounter the end of one of the iterators
end
```

Hopefully, it clarifies things. I'm currently working on a similar idea, but with one more iterator type (`PrefetchIterator`) instead of custom `safenext` and `complete`, which should resolve issues with type stability.